types/src/transaction/authenticator.rs

```diff
@@ -17,19 +20,112 @@ use rand::{rngs::OsRng, Rng};
+/// the transaction hash is well-formed and whether the sha3 hash of the
+/// `AccountAuthenticator`'s `AuthenticationKeyPreimage` matches the `AuthenticationKey` stored
+/// under the participating signer's account address.
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum TransactionAuthenticator {
+    /// Single-agent transaction.
+    SingleAgent { sender: AccountAuthenticator },
+    /// Multi-agent transaction.
+    MultiAgent {
+        sender: AccountAuthenticator,
+        secondary_signers: Vec<AccountAuthenticator>,
+    },
+}
+
+impl TransactionAuthenticator {
+    pub fn new_single_agent(sender: AccountAuthenticator) -> Self {
+        Self::SingleAgent { sender }
+    }
+
+    pub fn new_multi_agent(
+        sender: AccountAuthenticator,
+        secondary_signers: Vec<AccountAuthenticator>,
+    ) -> Self {
+        Self::MultiAgent {
+            sender,
+            secondary_signers,
+        }
+    }
+
+    /// Return Ok if all AccountAuthenticator's public keys match their signatures, Err otherwise
+    pub fn verify<T: Serialize + CryptoHash>(&self, message: &T) -> Result<()> {
+        match self {
+            Self::SingleAgent { sender } => sender.verify(message),
+            Self::MultiAgent {
+                sender,
+                secondary_signers,
+            } => {
+                sender.verify(message)?;
+                for signer in secondary_signers {
+                    signer.verify(message)?;
+                }
+                Ok(())
+            }
+        }
+    }
+
+    pub fn sender(&self) -> AccountAuthenticator {
+        match self {
+            Self::SingleAgent { sender } => sender.clone(),
+            Self::MultiAgent {
+                sender,
+                secondary_signers: _,
+            } => sender.clone(),
+        }
+    }
+
+    pub fn secondary_signers(&self) -> Vec<AccountAuthenticator> {
+        match self {
+            Self::SingleAgent { sender } => vec![],
+            Self::MultiAgent {
+                sender: _,
+                secondary_signers,
+            } => secondary_signers.to_vec(),
+        }
+    }
+}
+
+impl fmt::Display for TransactionAuthenticator {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::SingleAgent { sender } => {
+                write!(
+                    f,
+                    "TransactionAuthenticator[scheme: SingleAgent, sender: {}]",
+                    self.sender()
+                )
+            },
+            Self::MultiAgent {
+                sender,
+                secondary_signers,
+            } => {
+                write!(
+                    f,
+                    "TransactionAuthenticator[\n\
+                        \tscheme: MultiAgent, \n\
+                        \tsender: {}\n\
+                        \tsecondary signers: {:?}]",
```

Comment:
Nit: it's probably better to not use the debug print here so that the formatting of elements in the secondary signers match the formatting of the sender. e.g., 

```rust
self.secondary_signer().map(|x| format!("{}", x)).join(", ")
```